### DISCUSSION TITLE
python-boto3 & python-botocore packages missing

### DISCUSSION THREAD

**Discussion by cwoollard** [Homebrew/discussions#5159] [Everyday usage] [answered]
### Output of `brew config`

```shell
brew install python-boto3
Warning: python-boto3 1.33.0 is already installed and up-to-date.
To reinstall 1.33.0, run:
  brew reinstall python-boto3
chriswo@AT3832 rfc % brew reinstall python-boto3
==> Fetching python-boto3
==> Downloading https://files.pythonhosted.org/packages/47/87/87aa0bedda7591bb7da730a415c2ba8520c80e0a5011648de5ede40aad9d/boto3-1.33.0.tar.gz
###################################################################################################################################################################### 100.0%
==> Reinstalling python-boto3 
Error: An exception occurred within a child process:
  FormulaUnavailableError: No available formula with the name "python-boto3". Did you mean python-brotli?
```

### Output of `brew doctor`

```shell
brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  python-jmespath
  python-botocore
  python-pbr
  python-attrs
  python-platformdirs
  python-boto3
  python-s3transfer
```

### Description of issue

Some of the python packages seem to have vanished.

python-botocore
python-boto3

Is this related to the python@3.12 update?

**Comment by SMillerDev** [comment]
It's a result of https://github.com/Homebrew/homebrew-core/issues/157500

**Reply by cwoollard** [reply]
That is really annoying, especially when this issue exists as well. 

Not everybody cares about using virtual environments. Yes, they are a good feature, but they also add unnecessary complexity.

```
pip3 install boto3
error: externally-managed-environment

× This environment is externally managed
─> To install Python packages system-wide, try brew install
    xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-brew-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip.
    
    If you wish to install a non-brew packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

**Reply by SMillerDev** [reply]
> Not everybody cares about using virtual environments. Yes, they are a good feature, but they also add unnecessary complexity.

Those are unfortunately mandated by the python project. You can probably leave feedback for them somewhere.

**Reply by ryanchapman** [reply]
Haven't seen anything that works around this other than

```
python3 -m pip install boto3 --break-system-packages
```

and yes, likely to break on the next `brew upgrade`

(I hope someone dislikes my suggestion so much that they post a better solution than "set up a venv")

**Reply by SMillerDev** [reply]
You can install with pipx (which will set up a venv for you )

**Reply by cwoollard** [reply]
I tried with Pipx. That just generated a bunch of errors. In the end, I just used the --break-system-packages flag. It sounded scary but worked better than any supposedly more graceful solution.

Also, it seems that more and more Python packages are being removed from Homebrew.

Honestly, I hope someone comes up with a better solution than this. It just feels broken at the moment. I suspect it is going to get worse before it gets better.

**Reply by cho-m** [reply, marked as answer]
> --break-system-packages flag. It sounded scary but worked better than any supposedly more graceful solution

The output of externally maintained message will inform users of when they might want to use this (and other alternatives) in next `python@3.12` update - https://github.com/Homebrew/homebrew-core/pull/165681

---

> Also, it seems that more and more Python packages are being removed from Homebrew.

Because these were never supposed to be packaged in homebrew/core. Most were added in last year and actually went against our documentation. The added formulae led to more problems: some mentioned in https://github.com/Homebrew/homebrew-core/issues/157500, others from user reported issues, and a general increase in maintainer burden.

---

If you need to use Homebrew's Python, don't want to use virtualenv, and need libs instead of bins so can't use `pipx`, then you can follow guidance from above PR, i.e.
```
You may restore the old behavior of pip by passing
the '--break-system-packages' flag to pip, or by adding
'break-system-packages = true' to your pip.conf file. The latter
will permanently disable this error.

If you disable this error, we STRONGLY recommend that you additionally
pass the '--user' flag to pip, or set 'user = true' in your pip.conf
file. Failure to do this can result in a broken Homebrew installation.
```

**Reply by cwoollard** [reply]
That is a good response. Very helpful. 

I think it is probably about time to uninstall all my versions of Python (inc. associated packages) and then reinstall things cleanly. That will likely remove the errors I have and might make things work as intended.

Thanks for the guidance.

**Comment by wscarano** [comment]
Thanks this fixed it:  python3 -m pip install boto3 --break-system-packages --user.
